core/src/main/java/com/linecorp/armeria/internal/server/RouteDecoratingService.java

```diff
@@ -201,6 +202,17 @@ public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exc
             return service.serve(ctx, req);
         }
 
+        public List<HttpService> serviceChain(ServiceRequestContext ctx) {
```

Comment:
Can we do the search without create a new list? We could move the type check logic (`.as()` calls) here from `DefaultServiceRequestContext` then we could return early as soon as we find the service.

Reply:
Good point. I wanted to reduce duplicate calls by caching `serviceChain` instead of calling `router.findAll()` every time `findService()` is called. There is a trade-off. Since `findService` is unlikely to be called often, the cache hit rate may be low.

Either way is good to me.

Reply:
I was rather talking about returning a list in this method. We could optimize this like the following:

```
list = router.findAll(...);
for (routed : list) {
    if (routed.isPresent()) {
        match = routed.value().decorator().as(type);
        if (match != null) {
            return match;
        }
    }
}

// Match not found
return null;
```

We could also optimize `findAll()` or add an optimized version that returns an iterator so that the search is done only as much as needed.